jni/lib/src/jni_object.dart

```diff
@@ -0,0 +1,143 @@
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'jni_bindings_generated.dart';
+import 'extensions.dart';
+import 'jni_class.dart';
+import 'jvalues.dart';
+
+part 'jni_object_methods_generated.dart';
+
+/// JniObject is a convenience wrapper around a JNI local object reference.
+///
+/// It holds the object, its associated associated jniEnv etc..
+/// It should be distroyed with [delete] method after done.
+///
+/// It's valid only in the thread it was created.
+/// When passing to code that might run in a different thread (eg: a callback),
+/// consider obtaining a global reference and reconstructing the object.
+class JniObject {
+  JClass _cls;
+  final JObject _obj;
+  final Pointer<JniEnv> _env;
+  JniObject.of(this._env, this._obj, this._cls);
+
+  JniObject.fromJObject(Pointer<JniEnv> env, JObject obj)
+      : _env = env,
+        _obj = obj,
+        _cls = nullptr;
+
+  /// Reconstructs a JniObject from [r]
+  ///
+  /// [r] still needs to be explicitly deleted when
+  /// it's no longer needed to construct any JniObjects.
+  JniObject.fromGlobalRef(Pointer<JniEnv> env, JniGlobalObjectRef r)
+      : _env = env,
+        _obj = env.NewLocalRef(r._obj),
+        _cls = env.NewLocalRef(r._cls);
+
+  /// Delete the local reference contained by this object.
+  ///
+  /// Do not use a JniObject after calling [delete].
```

Comment:
Add a `boolean _deleted = false;` to `class JniObject`. Set it to true here. And check in all methods that it is false before using it.